### PR TITLE
feat: add support for gradient checkpointing for TimmWrapperModel and TimmWrapperForImageClassification

### PR DESCRIPTION
**Pull request by Yozer**
# What does this PR do?

This enables gradient checkpointing for timm wrapped models.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@qubvel

### CODE REVIEW

**Comment by qubvel** [comment]
cc @rwightman

**Comment by Yozer** [comment]
bump @qubvel @rwightman 
I'm happy to do any fixes if needed. Gradient checkpointing might help others train larger timm models on vram constrained systems like mine.

**Comment by rwightman** [comment]
I think it looks reasonable, might be some corner cases in terms of models that don't support gradient checkpointing asserting and thus throwing an exception, so that might need handling for tests? I feel it's fine for that to happen if someone tries to use the model and it's not supported.

So, in my tests this is used as the set_grad_checkpointing() fn can exist for a model but will assert & throw if not supported

```
    try:
        model.set_grad_checkpointing()
    except:
        # throws if not supported, that's fine
        pass
    else:
        # grad checkpointing supported
```

File: src/transformers/models/timm_wrapper/modeling_timm_wrapper.py [hunk 1]
```diff
@@ -118,8 +119,13 @@ def __init__(self, config: TimmWrapperConfig):
         # using num_classes=0 to avoid creating classification head
         extra_init_kwargs = config.model_args or {}
         self.timm_model = timm.create_model(config.architecture, pretrained=False, num_classes=0, **extra_init_kwargs)
+        if hasattr(self.timm_model, "set_grad_checkpointing"):
+            self.supports_gradient_checkpointing = True
```
Comment (qubvel) [inline src/transformers/models/timm_wrapper/modeling_timm_wrapper.py]: Let's define it for `TimmWrapperPreTrainedModel` model to avoid adding for both models

```py
        @property
        def supports_gradient_checkpointing(self):
            return hasattr(self.timm_model, "set_grad_checkpointing")
```

File: src/transformers/models/timm_wrapper/modeling_timm_wrapper.py [hunk 2]
```diff
@@ -118,8 +119,13 @@ def __init__(self, config: TimmWrapperConfig):
...
         self.post_init()
 
+    def _set_gradient_checkpointing(self, enable: bool = True, gradient_checkpointing_func: Callable = checkpoint):
+        self.timm_model.set_grad_checkpointing(enable)
```
Comment (qubvel) [inline src/transformers/models/timm_wrapper/modeling_timm_wrapper.py]: Let's also move to `TimmWrapperPreTrainedModel`

**Comment by Yozer** [comment]
I am checking if ```set_grad_checkpointing``` exists on underlying timm model but you're right, I can see in timm source code that sometimes set_grad_checkpointing has an assert.

Currently HF will throw if you try to use gradient checkpointing on model that doesn't support it so I would say that it's fine to throw from timm as well.

@rwightman  Do you mean any tests in particular? The one that were run are green, unless you meant a new test case for a model that can assert?

**Review by qubvel — COMMENTED** [review]
@Yozer Thanks for adding the feature, a few comments on my side:

**Comment by rwightman** [comment]
@Yozer I agree that if it throws when using an unsupported feature that's fine, the user will get the message and not try to use it... what I meant for tests is that if there are any tests for the feature, and they assume that based on the 'supports_gradient_checkpointing' flag being true, that grad checkpointing will work ... that won't be the case for timm models so the tests should be setup such that the failure is expected.

File: src/transformers/models/timm_wrapper/modeling_timm_wrapper.py (hunk 1 above)
Comment (Yozer) [reply src/transformers/models/timm_wrapper/modeling_timm_wrapper.py]: I had problems with @property, it wasn't overriding the base class variable.
I just used post_init to check support for timm model

File: src/transformers/models/timm_wrapper/modeling_timm_wrapper.py (hunk 2 above)
Comment (Yozer) [reply src/transformers/models/timm_wrapper/modeling_timm_wrapper.py]: done, all in base class

**Comment by Yozer** [comment]
@rwightman @qubvel 
Thanks for comments. Could you check the new approach?
1. Logic moved to base class
1. Added test for a model that doesn't support gradient checkpointing (it does the assert thing)
1. I did add try-except with AssertionError so ```supports_gradient_checkpointing``` should truly reflect if timm model supports it or no. I'm slightly afraid of doing enable/disable although I'm doing it very early in ```post_init``` so ```_backward_compatibility_gradient_checkpointing``` should work fine (and the tests for it are green)

File: src/transformers/models/timm_wrapper/modeling_timm_wrapper.py [hunk 3]
```diff
@@ -107,6 +112,24 @@ def _init_weights(self, module):
...
+        """
+        if not hasattr(self.timm_model, "set_grad_checkpointing"):
+            return False
+
+        try:
+            self.timm_model.set_grad_checkpointing(enable=True)
+            self.timm_model.set_grad_checkpointing(enable=False)
+            return True
+        except AssertionError:
+            return False
```
Comment (qubvel) [inline src/transformers/models/timm_wrapper/modeling_timm_wrapper.py]:
```suggestion
        try:
            self.timm_model.set_grad_checkpointing(enable=True)
            self.timm_model.set_grad_checkpointing(enable=False)
            return True
        except Exception:
            return False
```
Otherwise, we might have `None`, but expected values are True/False

File: src/transformers/models/timm_wrapper/modeling_timm_wrapper.py [hunk 4]
```diff
@@ -107,6 +112,24 @@ def _init_weights(self, module):
...
+
+    def _set_gradient_checkpointing(self, enable: bool = True, gradient_checkpointing_func: Callable = checkpoint):
```
Comment (qubvel) [inline src/transformers/models/timm_wrapper/modeling_timm_wrapper.py]:
```suggestion
    def _set_gradient_checkpointing(self, enable: bool = True, *args, **kwargs):
```

not used, so let's make it implicit + remove unnecessary import

File: src/transformers/models/timm_wrapper/modeling_timm_wrapper.py [hunk 5]
```diff
@@ -14,11 +14,12 @@
 # limitations under the License.
 
 from dataclasses import dataclass
-from typing import Optional, Union
+from typing import Callable, Optional, Union
 
 import torch
 from torch import Tensor, nn
 from torch.nn import BCEWithLogitsLoss, CrossEntropyLoss, MSELoss
+from torch.utils.checkpoint import checkpoint
```
Comment (qubvel) [inline src/transformers/models/timm_wrapper/modeling_timm_wrapper.py]:
```suggestion
```

File: src/transformers/models/timm_wrapper/modeling_timm_wrapper.py [hunk 6]
```diff
@@ -14,11 +14,12 @@
 # limitations under the License.
 
 from dataclasses import dataclass
-from typing import Optional, Union
+from typing import Callable, Optional, Union
```
Comment (qubvel) [inline src/transformers/models/timm_wrapper/modeling_timm_wrapper.py]:
```suggestion
from typing import Optional, Union
```

**Review by qubvel — COMMENTED** [review]
Thanks for the fix, please see the comments

**Comment by Yozer** [comment]
Done & done, thanks.
Ready for review @qubvel @rwightman

**Comment by Yozer** [comment]
@qubvel  @rwightman  bump

**Comment by rwightman** [comment]
@Yozer I think it looks fine from timm perspective, the state toggle should catch the models that don't support it, and it's lightweight enough and I feel it's pretty low risk (it toggles a bool, or bool per stage in some models). @qubvel is final review re Transformers side

**Review by qubvel — APPROVED** [review]
